Add docs for controllable parameters API (#73734)
metabase/metabase: docs/embedding/parameters.md

---
title: Modular embedding parameters
summary: Pass parameter values to embedded dashboards and SQL questions in modular embedding (React SDK and web components).
---

# Modular embedding parameters

This page covers how to pass parameter values to embedded dashboards and SQL questions.

## Modular embedding SDK (React)

### Pass parameter values to a dashboard

> You can set initial values for embeds (uncontrolled), and optionally keep your app in sync with values as people change them (controlled).

Reviewer: ```suggestion
You can set initial values for embeds (uncontrolled), and optionally keep your app in sync with values as people change them (controlled). You can pick either `initialParameters` and `parameters`, but don't combine them.
```

#### `initialParameters` (uncontrolled)

Set the filter values once on load. Your app won't know when people change filters in the dashboard. Pick this when you don't need to track those changes.

```typescript
{% include_file "{{ dirname }}/snippets/parameters/dashboards/initial-parameters.tsx" snippet="example" %}
```

See [How parameter values are resolved](#how-parameter-values-are-resolved) for null / missing-slug semantics.

#### `parameters` + `onParametersChange` (controlled)

> Push values from your app, and observe every applied change via `onParametersChange`. This works like a controlled `<input value onChange>` - your app holds the source of truth, the dashboard re-renders when the prop changes, and you receive a callback whenever applied values change.

Reviewer: ```suggestion
Push values from your app, and observe every applied change via `onParametersChange`. This works like a controlled `<input value onChange>`. Your app holds the source of truth, the dashboard re-renders when the prop changes, and you receive a callback whenever applied values change.
```

```typescript
{% include_file "{{ dirname }}/snippets/parameters/dashboards/controlled-parameters.tsx" snippet="example-controlled" %}
```

`onParametersChange` receives the [dashboard parameter change payload](#dashboard-parameter-change-payload).

> 
> 
> > Don't combine `initialParameters` and `parameters` - pick one. For controlled behavior, only use `parameters`.

Reviewer: ```suggestion
```
Included above.

### Pass parameters to SQL questions

> You can pass [parameter](../questions/native-editor/sql-parameters.md) values to SQL questions in the format `{parameter_name: parameter_value}`. You can set initial values for embeds (uncontrolled), and optionally keep your app in sync with values as people change them (controlled).

Reviewer: ```suggestion
You can pass [parameter](../questions/native-editor/sql-parameters.md) values to SQL questions in the format `{parameter_name: parameter_value}`. You can set initial values for embeds (uncontrolled), and optionally keep your app in sync with values as people change them (controlled).  You can pick either `initialSqlParameters` and `sqlParameters`. Don't combine them.
```

These props only work with SQL questions, not query-builder questions.

#### `initialSqlParameters` (uncontrolled)

Set the parameter values once on load. Your app won't know when people change parameters in the question. Pick this when you don't need to track those changes.

```typescript
{% include_file "{{ dirname }}/snippets/parameters/questions/initial-sql-parameters.tsx" snippet="example" %}
```

See [How parameter values are resolved](#how-parameter-values-are-resolved) for null / missing-slug semantics.

#### `sqlParameters` + `onSqlParametersChange` (controlled)

Push values from your app, and observe every applied change via `onSqlParametersChange`. This works like a controlled `<input value onChange>` - your app holds the source of truth, the question re-renders when the prop changes, and you receive a callback whenever applied values change.

```typescript
{% include_file "{{ dirname }}/snippets/parameters/questions/controlled-sql-parameters.tsx" snippet="example-controlled" %}
```

`onSqlParametersChange` receives the [SQL question parameter change payload](#sql-question-parameter-change-payload).

> > Don't combine `initialSqlParameters` and `sqlParameters` - pick one. For controlled behavior, only use `sqlParameters`.

Reviewer: ```suggestion
```

```html
<metabase-dashboard id="my-dashboard" dashboard-id="1"></metabase-dashboard>

<script>
  const el = document.getElementById("my-dashboard");

  el.addEventListener("parameters-change", (event) => {
    const { source, parameters, defaultParameters, lastUsedParameters } =
      event.detail;
    console.log(source, parameters);
  });
</script>
```

The `event.detail` carries the [dashboard parameter change payload](#dashboard-parameter-change-payload).

For SQL questions, listen for `sql-parameters-change` on `<metabase-question>`. Its `event.detail` carries the [SQL question parameter change payload](#sql-question-parameter-change-payload).

## How parameter values are resolved

These rules apply to all four props — `initialParameters` / `parameters` (dashboards) and `initialSqlParameters` / `sqlParameters` (SQL questions) — and to the matching web component attributes (`initial-parameters`, `parameters`, etc.). For each parameter slug:

- set to a value: that value is applied. For dashboards, use a `string` for a single-option filter and an array of `string`s for multi-option filters.
- set to `null`: the parameter is cleared and its default is not used.

> - omitted (or set to `undefined`): falls back to the parameter's default (or `null` if it has no default).

Reviewer: ```suggestion
- **Set a value**: Pass a `string` for a single-option filter, and an array of `string`s for multi-option filters.
- **Clear a value:** Set to `null`: the parameter is cleared and its default is not used.
- **Reset to the default value**: Omit a value (or set to `undefined`) and the embed will fall back to the parameter's default (or `null` if it has no default).
```
In that last one, would an empty object work?

---
title: Modular embedding parameters
summary: Pass parameter values to embedded dashboards and SQL questions in modular embedding (React SDK and web components).
---

# Modular embedding parameters

> This page covers how to pass parameter values to embedded dashboards and SQL questions in [modular embedding](./modular-embedding.md), for both the [Modular embedding SDK](./sdk/introduction.md) (React) and the web components.

Reviewer: ```suggestion
This page covers how to pass parameter values to embedded dashboards and SQL questions.
```

## Modular embedding SDK (React)

### Pass parameter values to a dashboard

> You can pre-populate dashboard filters from your app, and (optionally) keep your app in sync with what's currently applied. Use one of these props depending on whether you only want to seed initial values or stay in sync over time.

Reviewer: ```suggestion
You can set initial values for embeds (uncontrolled), and optionally keep your app in sync with values as people change them (controlled).
```

#### `initialParameters` (uncontrolled)

Pre-populate filters once on load. After the dashboard mounts, user widget edits are not reflected back to your app. Pick this when you don't need to react to filter changes.

```typescript
{% include_file "{{ dirname }}/snippets/parameters/dashboards/initial-parameters.tsx" snippet="example" %}
```

> For each parameter:

Reviewer: Can we consolidate this section? It's identical to the Uncontrolled section.

- set to a value (a `string` for a single option, an array of `string`s for multiple): that value is applied.
- set to `null`: strictly cleared, ignoring the parameter's default.
- omitted (or set to `undefined`): falls back to the parameter's default (or `null` if it has no default).

Examples:

> - Passing an empty object `{}` resets every parameter to its default.

Reviewer: Above we said "omit" or set to undefined to reset to default. But here we're saying to pass an empty object?

Do we even need these examples? Should the be inlined in the bullets above?

Author: Let's remove the examples

- Passing an object with every slug set to `null` clears all parameter values.

#### `parameters` + `onParametersChange` (controlled)

Push values from your app, and observe every applied change via `onParametersChange`. This works like a controlled `<input value onChange>` - your app holds the source of truth, the dashboard re-renders when the prop changes, and you receive a callback whenever applied values change.

```typescript
{% include_file "{{ dirname }}/snippets/parameters/dashboards/controlled-parameters.tsx" snippet="example-controlled" %}
```

`onParametersChange` receives the [dashboard parameter change payload](#dashboard-parameter-change-payload).

> ##### Clearing a single parameter

Reviewer: Do we need sections and sample code for clearing params? That's a lot of ceremony for a simple value to pass to clear params.

Author: Let's remove, we have `How parameter values are resolved`

Set its value to `null`. This strictly clears the parameter and ignores its default value. Missing slugs fall back to `parameter.default ?? null`.

```typescript
{% include_file "{{ dirname }}/snippets/parameters/dashboards/controlled-parameters.tsx" snippet="example-clear-one" %}
```

##### Clearing all parameters

Pass an empty object.

```typescript
{% include_file "{{ dirname }}/snippets/parameters/dashboards/controlled-parameters.tsx" snippet="example-clear-all" %}
```

To switch back to uncontrolled mode (and leave the dashboard's last applied values in place), set the `parameters` prop to `undefined`.

> Don't combine `initialParameters` and `parameters` - pick one. For controlled behavior, only use `parameters`.

### Pass parameters to SQL questions

You can pass [parameter](../questions/native-editor/sql-parameters.md) values to SQL questions like so:
```
{parameter_name: parameter_value}

```

> There are two ways to pass parameters:

Reviewer: We can probably do a table or two to put all this stuff together. Someting like:

**SDK**

| Embed         | Uncontrolled            | Controlled                                  |
| ------------- | ----------------------- | ------------------------------------------- |
| Dashboard     | `initialParameters`     | `parameters` + `onParametersChange`         |
| SQL question  | `initialSqlParameters`  | `sqlParameters` + `onSqlParametersChange`   |

**Web components**

| Embed         | Uncontrolled (attribute)  | Controlled (property + event)               |
| ------------- | ------------------------- | ------------------------------------------- |
| Dashboard     | `initial-parameters`      | `.parameters` + `parameters-change`         |
| SQL question  | `initial-sql-parameters`  | `.sqlParameters` + `sql-parameters-change`  |
